File: tools/sync-pob-catalog/src/extract_stat_descriptions.lua
-- extract_stat_descriptions.lua — sync-pob-catalog `extract-lua --what stat-descriptions`
--
-- 在**最小环境**（描述文件是纯数据 `return {...}`，无函数/flag，loadfile 即可）下
-- 加载 vendor/PathOfBuilding-PoE2 的 Data/StatDescriptions/<scope>.lua（+ 沿 `parent`
-- 链继承的 root），为每个 stat_id 渲染一段 canonical 文本（喂代表值 V=1），把
-- stat_id → 文本输出为 JSONL（每行一个 JSON 对象）到 stdout：
--
--   {"stat":"<id>","scope":"<scope>","text":"<rendered line>","line":N,"compound":false}
--   {"stat":"<id>","scope":"<scope>","compound":true,"member_stats":[...],"template":"<raw text>"}
--
-- 渲染逻辑忠实移植自 vendor Modules/StatDescriber.lua（matchLimit:41-61 选变体 +
-- gsub 链:281-315 替换占位符 + 按字面 `\n` 切行:316）。**省略 applySpecial 值变换**
-- （negate/divide_by_* 只改数值不改词形；模板提取只需词形——ModName/ModType 由 §B
-- 解析得出，运行期填真实掷值）。**多 stat（compound）不强解**，原样输出供 §B 处置。
--
-- 确定性约定：本脚本只忠实抽取 + 合法 JSON；排序与 byte-stable 序列化由 Rust 侧。
--
-- 用法（由 Rust 侧经 stdin 注入执行）：
--   luajit - <vendor_src_dir> <scope_file>   -- scope_file 不含 .lua（如 passive_skill_stat_descriptions）

local vendorSrc = arg and arg[1]
local scopeFile = arg and arg[2]
if not vendorSrc or not scopeFile then
	io.stderr:write("usage: luajit - <vendor_src_dir> <scope_file_without_.lua>\n")
	os.exit(2)
end

----------------------------------------------------------------------
-- JSON 编码（最小，仅本脚本所需的 string/number/bool/array 形态）。
----------------------------------------------------------------------
local function jsonEscape(s)
	return (s:gsub('[%z\1-\31\\"]', function(c)
		if c == '"' then return '\\"'
		elseif c == "\\" then return "\\\\"
		elseif c == "\n" then return "\\n"
		elseif c == "\r" then return "\\r"
		elseif c == "\t" then return "\\t"
		else return string.format("\\u%04x", string.byte(c)) end
	end))
end
local function jsonStr(s) return '"' .. jsonEscape(s) .. '"' end
local function jsonArr(t)
	local parts = {}
	for _, v in ipairs(t) do parts[#parts + 1] = jsonStr(v) end
	return "[" .. table.concat(parts, ",") .. "]"
end

----------------------------------------------------------------------
-- 加载一个 scope 描述文件（纯数据）。沿 `parent` 链返回 scopeList（child 在前）。
----------------------------------------------------------------------
local loaded = {}
local function loadScope(name)
	if loaded[name] then return loaded[name] end
	local path = vendorSrc .. "/Data/StatDescriptions/" .. name .. ".lua"
	local chunk, err = loadfile(path)
	if not chunk then
		io.stderr:write("loadfile failed: " .. tostring(err) .. "\n")
		os.exit(3)
	end
	local scope = chunk()
	scope.__name = name
	if scope.parent then
		scope.__scopeList = { scope }
		for _, s in ipairs(loadScope(scope.parent)) do
			scope.__scopeList[#scope.__scopeList + 1] = s
		end
	else
		scope.__scopeList = { scope }
	end
	loaded[name] = scope.__scopeList
	return scope.__scopeList
end

----------------------------------------------------------------------
-- matchLimit（移植自 StatDescriber.lua:41-61，单值 quality=nil）：按 val 选首个命中
-- limit 的变体。val[i] = {min=V,max=V}。
----------------------------------------------------------------------
local function matchLimit(lang, val)
	for _, desc in ipairs(lang) do
		if not desc.gem_quality then
			local match = true
			for i, limit in ipairs(desc.limit) do
				local v = val[i] or { min = 0, max = 0 }
				if limit[1] == "!" then
					if v.min == limit[2] then match = false break end
				elseif (limit[2] ~= "#" and v.min > limit[2]) or (limit[1] ~= "#" and v.min < limit[1]) then
					match = false break
				end
			end
			if match then return desc end
		end
	end
end

----------------------------------------------------------------------
-- 渲染变体文本（移植自 StatDescriber.lua:281-315 gsub 链；省 applySpecial）。
-- 所有占位符代入 val[n].min（= 代表值），单值无区间。
----------------------------------------------------------------------
local function renderText(text, val)
	local function single(i)
		local v = val[i] or { min = 0 }
		return tostring(v.min)
	end
	return (text
		:gsub("{(%d)}", function(n) return single(tonumber(n) + 1) end)
		:gsub("{}", function() return single(1) end)
		:gsub("{:%+?d}", function() local v = val[1] or { min = 0 }; if v.min >= 0 then return "+" .. v.min else return tostring(v.min) end end)
		:gsub("{(%d):(%+?)d}", function(n, fmt)
			local v = val[tonumber(n) + 1] or { min = 0 }
			if fmt == "+" and v.min >= 0 then return "+" .. v.min else return tostring(v.min) end
		end)
		:gsub("%%%%", "%%"))
end

-- 按字面 `\n`（数据里是 `\\n`）切行（移植自 :316）。
local function splitLines(s)
	local out = {}
	for line in (s .. "\\n"):gmatch("([^\\]+)\\n") do out[#out + 1] = line end
	return out
end

----------------------------------------------------------------------
-- 为一个 descriptor + 代表值 V 渲染 canonical 文本行。返回 lines 或 nil。
----------------------------------------------------------------------
local function describe(descriptor, v)
	local lang = descriptor[1]
	if not lang then return nil end
	local val = {}
	for i = 1, #descriptor.stats do val[i] = { min = v, max = v } end
	local desc = matchLimit(lang, val)
	if not desc then return nil end
	return splitLines(renderText(desc.text, val))
end

----------------------------------------------------------------------
-- 主：对每个 scope（逗号分隔）遍历其 lookup 表（`["stat_id"]=idx`），逐 stat 渲染。
-- 多 scope 时各自只抽自己的 own keys（不抽继承的 parent，避免重复——root 单独抽）；
-- 跨 scope 同 stat_id 由 Rust 侧按 scope 精确度去重。
----------------------------------------------------------------------
local function processScope(scopeFile)
local scopeList = loadScope(scopeFile)
local scope = scopeList[1]
local emitted = {}

-- lookup 键 = scope 里的字符串键（排除 parent/name/__* + 整数键 descriptor）。
local statIds = {}
for k, v in pairs(scope) do
	if type(k) == "string" and type(v) == "number" and not k:match("^__") and k ~= "parent" and k ~= "name" then
		statIds[#statIds + 1] = k
	end
end
table.sort(statIds)

for _, statId in ipairs(statIds) do
	if not emitted[statId] then
		emitted[statId] = true
		local descriptor = scope[scope[statId]]
		if descriptor and descriptor.stats then
			if #descriptor.stats == 1 then
				-- 单 stat：喂 V=1 渲染；若无命中变体（仅负向）退 V=-1。
				local lines = describe(descriptor, 1) or describe(descriptor, -1)
				if lines then
					for i, line in ipairs(lines) do
						io.write(string.format(
							'{"stat":%s,"scope":%s,"text":%s,"line":%d,"compound":false}\n',
							jsonStr(statId), jsonStr(scopeFile), jsonStr(line), i))
					end
				else
					io.write(string.format('{"stat":%s,"scope":%s,"compound":false,"unrendered":true}\n',
						jsonStr(statId), jsonStr(scopeFile)))
				end
			else
				-- compound（多 stat）：原样输出 raw text + member_stats，不强解。
				-- 多个 member stat_id 共指同一 descriptor——标记全部 member，避免重复输出。
				local members = {}
				for _, s in ipairs(descriptor.stats) do
					members[#members + 1] = s
					emitted[s] = true
				end
				local lang = descriptor[1]
				local rawText = (lang and lang[1] and lang[1].text) or ""
				io.write(string.format(
					'{"stat":%s,"scope":%s,"compound":true,"member_stats":%s,"template":%s}\n',
					jsonStr(descriptor.stats[1]), jsonStr(scopeFile), jsonArr(members), jsonStr(rawText)))
			end
		end
	end
end
end

-- 分发：arg[2] 是逗号分隔的 scope 列表（对齐 sync-pob-catalog `--files` 约定）。
for scope in (scopeFile .. ","):gmatch("([^,]+),") do
	processScope((scope:gsub("%s+", "")))
end

File: tools/sync-pob-catalog/src/extract_stat_descriptions.rs
//! `extract-lua --what stat-descriptions`：vendor PoB2
//! `Data/StatDescriptions/*.lua` → 每个 stat_id 的 canonical 显示文本 →
//! `data/<版本>/overlay/stat_descriptions.json`（「stat_id → Modifier
//! 第二通道」的数据面）。
//!
//! 职责切分与 [`crate::extract_stat_map`] 一致：Lua 引导脚本
//! （`extract_stat_descriptions.lua`，编译期内嵌）在最小环境下加载描述表、对
//! 每个 stat_id 喂代表值（V=1）渲染文本，以 JSONL 输出（每行一条 single 文本
//! 行或一条 compound 模板）；Rust 侧统一做 scope 分段、行序归并（[`BTreeMap`]
//! 字典序）与整体序列化，保证同输入重跑 **byte-stable**。
//!
//! 消费侧 schema 单一来源 = [`pobr_data::catalog::stat_descriptions`]（生成 /
//! 消费两侧共用同一 serde 形状，防字段漂移）；precedence（child scope 覆盖
//! parent）与「哪条 stat_id 受支持」的判定是消费侧 §B 生成器 / `parse_mod_engine`
//! 的职责，本模块不做。

use std::collections::BTreeMap;
use std::io;

use pobr_data::catalog::stat_descriptions::{CompoundDescription, StatDescriptionsDef};
use serde::{Deserialize, Serialize};

use crate::extract_lua::{
    ExtractLuaArgs, OverlayMeta, invoke_luajit_jsonl, read_vendor_version, resolve_version_file,
};

/// 引导脚本内容（经 stdin 注入 luajit，二进制自包含、不依赖运行目录）。
const BOOTSTRAP_LUA: &str = include_str!("extract_stat_descriptions.lua");

/// 当前 overlay 文档 schema 标识（字段演化时递增）。
pub const STAT_DESCRIPTIONS_SCHEMA: &str = "stat_descriptions/v1";

/// 默认抽取 scope（tree 通道最相关：root + passive + presence/aura）。
/// 其余 StatDescriptions 文件（skill/gem/monster/advanced_mod）按需经 `--files` 加。
pub const DEFAULT_STAT_DESC_FILES: &[&str] = &[
    "stat_descriptions",
    "passive_skill_stat_descriptions",
    "passive_skill_aura_stat_descriptions",
];

/// 引导脚本输出的单行 JSONL：一条 single 文本行 或 一条 compound 模板。
///
/// - single 渲染：`{stat, scope, text, line, compound:false}`
/// - single 无变体：`{stat, scope, compound:false, unrendered:true}`
/// - compound：`{stat, scope, compound:true, member_stats, template}`
#[derive(Debug, Clone, Deserialize)]
pub struct StatDescRow {
    /// stat 稳定 id（compound 行 = descriptor 首个 stat_id）。
    pub stat: String,
    /// 来源 scope 名（StatDescriptions 文件名，不含 `.lua`）。
    pub scope: String,
    /// 是否多 stat（compound）描述符。
    pub compound: bool,
    /// 渲染出的单行文本（仅 single 且可渲染）。
    #[serde(default)]
    pub text: Option<String>,
    /// 行序号（1-based；同一 stat 的多行描述按此排序）。
    #[serde(default)]
    pub line: Option<u32>,
    /// 无可渲染变体标记（仅 single）。
    #[serde(default)]
    pub unrendered: bool,
    /// compound 绑定的全部 stat_id（仅 compound）。
    #[serde(default)]
    pub member_stats: Vec<String>,
    /// compound 原样模板（仅 compound）。
    #[serde(default)]
    pub template: Option<String>,
}

/// 完整 overlay 文档（生成侧；`_meta` 头部 + 消费侧 [`StatDescriptionsDef`] 平铺）。
#[derive(Debug, Serialize, Deserialize)]
pub struct StatDescriptionsDoc {
    /// 头部元信息（serde 落为 `_meta`，置于文件最前）。
    #[serde(rename = "_meta")]
    pub meta: OverlayMeta,
    /// 描述本体（按 scope 分段，BTreeMap 字典序保证确定性）。
    #[serde(flatten)]
    pub def: StatDescriptionsDef,
}

/// 执行抽取，返回最终（byte-stable 的）JSON 文本。
pub fn run_extract_stat_descriptions(args: &ExtractLuaArgs) -> io::Result<String> {
    let rows = invoke_luajit_jsonl::<StatDescRow>(args, BOOTSTRAP_LUA)?;
    let meta = build_meta(args)?;
    assemble_stat_descriptions_document(meta, rows)
}

/// 组装最终文档：JSONL 行 → 各 scope 的 single / compound / unrendered 三段。
///
/// single 文本行先归并到 `BTreeMap<line, text>`（行序确定，与输入次序无关），再
/// 展平为 `Vec<String>`——保证同输入 byte-stable。重复键（同 scope 同 stat 同
/// line，或 compound 重复）报错不静默覆盖。
pub fn assemble_stat_descriptions_document(
    meta: OverlayMeta,
    rows: Vec<StatDescRow>,
) -> io::Result<String> {
    // 中间态：scope → stat → line → text（行序归并）。
    let mut single_lines: BTreeMap<String, BTreeMap<String, BTreeMap<u32, String>>> =
        BTreeMap::new();
    let mut def = StatDescriptionsDef::default();

    for row in rows {
        let scope = def.scopes.entry(row.scope.clone()).or_default();
        if row.compound {
            let template = row.template.clone().ok_or_else(|| {
                io::Error::other(format!(
                    "stat-descriptions compound 行缺 template：{}",
                    row.stat
                ))
            })?;
            let entry = CompoundDescription {
                member_stats: row.member_stats.clone(),
                template,
            };
            if scope.compound.insert(row.stat.clone(), entry).is_some() {
                return Err(io::Error::other(format!(
                    "stat-descriptions 抽取重复 compound 键：{}::{}",
                    row.scope, row.stat
                )));
            }
        } else if row.unrendered {
            scope.unrendered.insert(row.stat.clone());
        } else {
            let text = row.text.clone().ok_or_else(|| {
                io::Error::other(format!(
                    "stat-descriptions single 行缺 text：{}::{}",
                    row.scope, row.stat
                ))
            })?;
            let line = row.line.unwrap_or(1);
            let slot = single_lines
                .entry(row.scope.clone())
                .or_default()
                .entry(row.stat.clone())
                .or_default();
            if slot.insert(line, text).is_some() {
                return Err(io::Error::other(format!(
                    "stat-descriptions 抽取重复 single 行：{}::{} line {line}",
                    row.scope, row.stat
                )));
            }
        }
    }

    // 行序归并展平到消费侧形状。
    for (scope_name, stats) in single_lines {
        let scope = def.scopes.entry(scope_name).or_default();
        for (stat, lines) in stats {
            scope.single.insert(stat, lines.into_values().collect());
        }
    }

    let doc = StatDescriptionsDoc { meta, def };
    let mut json = serde_json::to_string_pretty(&doc).map_err(io::Error::other)?;
    json.push('\n');
    Ok(json)
}

/// 读取 vendor 版本文件并构建 `_meta`（与 `extract_lua::build_meta` 同约定）。
fn build_meta(args: &ExtractLuaArgs) -> io::Result<OverlayMeta> {
    let (commit, subject) = read_vendor_version(&resolve_version_file(args))?;

    let extracted_files = args
        .files
        .iter()
        .map(|name| format!("Data/StatDescriptions/{name}.lua"))
        .collect();

    let mut regen = format!(
        "cargo run -p sync-pob-catalog -- extract-lua --what stat-descriptions --vendor-root vendor/PathOfBuilding-PoE2/src --files {}",
        args.files.join(",")
    );
    if let Some(out) = &args.out_for_meta {
        regen.push_str(&format!(" --out {out}"));
    }

    Ok(OverlayMeta {
        schema: STAT_DESCRIPTIONS_SCHEMA.to_string(),
        generator: "sync-pob-catalog extract-lua".to_string(),
        vendor: "PathOfBuilding-PoE2".to_string(),
        vendor_commit: commit,
        vendor_commit_subject: subject,
        extracted_files,
        regen_command: regen,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn meta() -> OverlayMeta {
        OverlayMeta {
            schema: STAT_DESCRIPTIONS_SCHEMA.to_string(),
            generator: "test".to_string(),
            vendor: "PathOfBuilding-PoE2".to_string(),
            vendor_commit: "0".repeat(40),
            vendor_commit_subject: "test".to_string(),
            extracted_files: vec![],
            regen_command: "test".to_string(),
        }
    }

    fn single(scope: &str, stat: &str, text: &str, line: u32) -> StatDescRow {
        StatDescRow {
            stat: stat.to_string(),
            scope: scope.to_string(),
            compound: false,
            text: Some(text.to_string()),
            line: Some(line),
            unrendered: false,
            member_stats: vec![],
            template: None,
        }
    }

    /// single 多行按 line 序展平；多 scope 各归各段；键字典序（BTreeMap 保证）。
    #[test]
    fn assembles_single_lines_in_order() {
        let json = assemble_stat_descriptions_document(
            meta(),
            vec![
                single(
                    "stat_descriptions",
                    "additional_strength",
                    "+1 to Strength",
                    1,
                ),
                // 故意乱序输入：line 2 在 line 1 之前。
                single("stat_descriptions", "two_line", "second", 2),
                single("stat_descriptions", "two_line", "first", 1),
                single(
                    "passive_skill_stat_descriptions",
                    "z_passive",
                    "passive text",
                    1,
                ),
            ],
        )
        .unwrap();
        let doc: StatDescriptionsDoc = serde_json::from_str(&json).unwrap();
        let root = &doc.def.scopes["stat_descriptions"];
        assert_eq!(root.single["additional_strength"], vec!["+1 to Strength"]);
        // 乱序输入仍按 line 序展平。
        assert_eq!(root.single["two_line"], vec!["first", "second"]);
        assert!(
            doc.def
                .scopes
                .contains_key("passive_skill_stat_descriptions")
        );
    }

    /// compound 行原样保留 template + member_stats。
    #[test]
    fn keeps_compound_template_verbatim() {
        let row = StatDescRow {
            stat: "a_stat".to_string(),
            scope: "stat_descriptions".to_string(),
            compound: true,
            text: None,
            line: None,
            unrendered: false,
            member_stats: vec!["a_stat".to_string(), "b_stat".to_string()],
            template: Some("Deal {0} damage to {1} targets".to_string()),
        };
        let json = assemble_stat_descriptions_document(meta(), vec![row]).unwrap();
        let doc: StatDescriptionsDoc = serde_json::from_str(&json).unwrap();
        let c = &doc.def.scopes["stat_descriptions"].compound["a_stat"];
        assert_eq!(c.member_stats, vec!["a_stat", "b_stat"]);
        assert_eq!(c.template, "Deal {0} damage to {1} targets");
    }

    /// unrendered 行进诊断集。
    #[test]
    fn records_unrendered_stats() {
        let row = StatDescRow {
            stat: "weird_stat".to_string(),
            scope: "stat_descriptions".to_string(),
            compound: false,
            text: None,
            line: None,
            unrendered: true,
            member_stats: vec![],
            template: None,
        };
        let json = assemble_stat_descriptions_document(meta(), vec![row]).unwrap();
        let doc: StatDescriptionsDoc = serde_json::from_str(&json).unwrap();
        assert!(
            doc.def.scopes["stat_descriptions"]
                .unrendered
                .contains("weird_stat")
        );
    }

    /// 重复 single 行（同 scope 同 stat 同 line）→ 报错不静默覆盖。
    #[test]
    fn duplicate_single_line_errors_out() {
        let result = assemble_stat_descriptions_document(
            meta(),
            vec![
                single("stat_descriptions", "dup", "a", 1),
                single("stat_descriptions", "dup", "b", 1),
            ],
        );
        assert!(result.is_err());
    }

    /// 字典序：a_stat 文本位置先于 z_stat（byte-stable 排序的可见证据）。
    #[test]
    fn output_is_lexicographically_sorted() {
        let json = assemble_stat_descriptions_document(
            meta(),
            vec![
                single("stat_descriptions", "z_stat", "z", 1),
                single("stat_descriptions", "a_stat", "a", 1),
            ],
        )
        .unwrap();
        assert!(json.find("a_stat").unwrap() < json.find("z_stat").unwrap());
    }
}
